### PR TITLE
dhdl plot for the TI

### PR DESCRIPTION
**Pull request by xiki-tempula**
Partially addresses #73

I have ported the dhdl plot for TI from the alchemical analysis.

Based on user experience, I have changed a few things.

- The user could now set custom label to the figure
- The x axis is elevated a bit to avoid bumping to the dhdl
- The user could now custom the color of the plot

### CODE REVIEW

**Comment by xiki-tempula** [comment]
@orbeckst I'm sorry that I felt that I cannot bump the test coverage to any higher.
I have ported the code from the alchemical analysis and modified it such that it works in py38.

However, there are two regions that I'm not quite sure of what it does.

```
    def getInd(r=ri, z=[0]):
        primo = r[0]
        min_dl = ndx * 0.02 * 2 ** (primo > 10)
        if dl_mat[primo].max() < min_dl:
            return z
        for i in r:
            for j in range(len(xs)):
                if dl_mat[i, j] > min_dl:
                    z.append(j)
                    return getInd(ri[j:], z)
```
This function exists to make sure that there won't be too many ticks in the x axis.
I have tested this function with
```
    ti_coul.dhdl = pd.DataFrame.from_dict(
        {'fep': range(100)},
        orient='index',
        columns=np.arange(100)/100).T
    assert isinstance(plot_ti_dhdl(ti_coul),
               matplotlib.axes.Axes)
```
but cannot get it fully covered.

The other part is
```
    if ndx > 1:
        lenticks = len(ax.get_ymajorticklabels()) - 1
        if min_y < 0: lenticks -= 1
        if lenticks < 5:
            from matplotlib.ticker import AutoMinorLocator as AML
            ax.yaxis.set_minor_locator(AML())
```
Which my understanding is to deal with the case when there are too tittle ticks in the x axis.
I have tested it with
```
    ti_coul.dhdl = pd.DataFrame.from_dict(
        {'fep': range(3)},
        orient='index',
        columns=np.arange(3)/3).T
    assert isinstance(plot_ti_dhdl(ti_coul),
               matplotlib.axes.Axes)
```
but cannot get it fully covered.

**Review by orbeckst — CHANGES_REQUESTED** [review]
Nice work and kudos for pushing up coverage as high as you did – it's hard with plots!

Given that you identified the code pieces that are hard to test, mark them explicitly as not included in coverage. Something like
```python
# pragma: no cover
```
You might have to find out how exactly to do this (we do it in MDAnalysis) but with this you explicitly acknowledge that certain code is untested and then it's removed from the coverage calculation.

**Comment by xiki-tempula** [comment]
@orbeckst Thank you for the advice. I have commented out the relevant lines with `# pragma: no cover`.

File: src/alchemlyb/visualisation/ti_dhdl.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        spine.set_zorder(12.2)
+
+    # Make level names
+    if labels is None:
+        lv_names2 = []
+        for dhdl in dhdl_data:
+            # Assume that the dhdl has only one columns
+            lv_names2.append(dhdl.dhdl.columns.values[0].capitalize())
+    else:
+        assert len(labels) == len(dhdl_data), \
```
Comment (orbeckst) [inline src/alchemlyb/visualisation/ti_dhdl.py:72]: Change this `assert` to a `raise ValueError`.

We should not be using assertions to capture user errors.

File: src/alchemlyb/visualisation/ti_dhdl.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+            'Length of labels ({}) should be the same as the number of data ({})'.format(len(labels), len(dhdl_data))
+        lv_names2 = labels
+
+    if colors is None:
+        colors = ['r', 'g', '#7F38EC', '#9F000F', 'b', 'y']
+    else:
+        assert len(colors) >= len(dhdl_data), \
```
Comment (orbeckst) [inline src/alchemlyb/visualisation/ti_dhdl.py:79]: `raise ValueError` instead

If this is not tested, use `#pragma: no cover`

File: src/alchemlyb/visualisation/ti_dhdl.py (hunk 1 above)
Comment (orbeckst) [reply src/alchemlyb/visualisation/ti_dhdl.py:72]: If this is not tested, use `#pragma: no cover`

**Review by orbeckst — CHANGES_REQUESTED** [review]
Thanks for adjusting coverage. I just saw two things to change: raise exceptions instead of using assertions for input checking – this was probably already in alchemical-analysis but should certainly not be in alchemlyb.

Otherwise, this looks excellent.

**Review by orbeckst — APPROVED** [review]

**Comment by orbeckst** [comment]
Thank you @xiki-tempula !! 🚀

**Pull request merged**